### PR TITLE
Added "Bouncer", which prohibits the creation of large projectiles.

### PR DESCRIPTION
**Pull request by AgaSpace**
I recently discovered a bug in the game with which we can increase projectiles. That would be okay, but it can cause very strong negative emotions on the server. So I took care of it and highlighted a few projectiles that we can increase.

### CODE REVIEW

**Review by hakusaro — DISMISSED** [review]

File: TShockAPI/Bouncer.cs [hunk 1]
```diff
@@ -2496,5 +2511,45 @@ internal static int GetMaxPlaceStyle(int tileID)
...
+			{ 425, 0 },
+			{ 426, 0 },
+
+			{ 612, 0 },
+			{ 953, 0 },
+
+			{ 756, 0 },
+			{ 522, 0 }
+		};
+		private Dictionary<short, short> Projectile_MaxValuesAI2 = new Dictionary<short, short> {
```
Comment (Arthri) [inline TShockAPI/Bouncer.cs]: I think these should be `<short, float>` dictionaries since AI[] is an array of floats

File: TShockAPI/Bouncer.cs [hunk 2]
```diff
@@ -2496,5 +2511,45 @@ internal static int GetMaxPlaceStyle(int tileID)
...
+			{ 405, 2 },
```
Comment (Arthri) [inline TShockAPI/Bouncer.cs]:
```suggestion
			{ 405, 1.2f },
```

Source: `Projectile.AI()`, `this.ai[1] = (float)Main.rand.Next(80, 121) / 100f;`

File: TShockAPI/Bouncer.cs [hunk 3]
```diff
@@ -2496,5 +2511,45 @@ internal static int GetMaxPlaceStyle(int tileID)
...
+			{ 950, 0 }
+		};
+		private Dictionary<short, short> Projectile_MaxValuesAI = new Dictionary<short, short> {
+			{ 611, 1 },
+
+			{ 950, 0 }
+		};
+
+		private Dictionary<short, short> Projectile_MinValuesAI2 = new Dictionary<short, short> {
+			{ 405, 0 },
```
Comment (Arthri) [inline TShockAPI/Bouncer.cs]:
```suggestion
			{ 405, 0.8f },
```

Source: `Projectile.AI()`, `this.ai[1] = (float)Main.rand.Next(80, 121) / 100f;`

File: TShockAPI/Bouncer.cs [hunk 4]
```diff
@@ -2496,5 +2511,45 @@ internal static int GetMaxPlaceStyle(int tileID)
...
+			{ 410, 0 },
```
Comment (Arthri) [inline TShockAPI/Bouncer.cs]:
```suggestion
			{ 410, 0.6f },
```
Source: `Projectile.AI()`, `this.ai[1] = (float)Main.rand.Next(60, 121) / 100f;`

File: TShockAPI/Bouncer.cs [hunk 5]
```diff
@@ -2496,5 +2511,45 @@ internal static int GetMaxPlaceStyle(int tileID)
...
+
+			{ 612, 0 },
+			{ 953, 0 },
+
+			{ 756, 0 },
+			{ 522, 0 }
+		};
+		private Dictionary<short, short> Projectile_MaxValuesAI2 = new Dictionary<short, short> {
+			{ 405, 2 },
+			{ 410, 2 },
```
Comment (Arthri) [inline TShockAPI/Bouncer.cs]:
```suggestion
			{ 410, 1.2f },
```
Source: `Projectile.AI()`, `this.ai[1] = (float)Main.rand.Next(60, 121) / 100f;`

File: TShockAPI/Bouncer.cs [hunk 6]
```diff
@@ -2496,5 +2511,45 @@ internal static int GetMaxPlaceStyle(int tileID)
...
+
+		private Dictionary<short, short> Projectile_MinValuesAI2 = new Dictionary<short, short> {
+			{ 405, 0 },
+			{ 410, 0 },
+
+			{ 424, 0 },
+			{ 425, 0 },
+			{ 426, 0 },
+
+			{ 612, 0 },
```
Comment (Arthri) [inline TShockAPI/Bouncer.cs]:
```suggestion
			{ 612, 0.4f },
```
Source: `Player.ItemCheck_Shoot()`, `float ai4 = (Main.rand.NextFloat() - 0.5f) * ((float)Math.PI / 4f);`

File: TShockAPI/Bouncer.cs [hunk 7]
```diff
@@ -2496,5 +2511,45 @@ internal static int GetMaxPlaceStyle(int tileID)
...
+		};
+		private Dictionary<short, short> Projectile_MaxValuesAI2 = new Dictionary<short, short> {
+			{ 405, 2 },
+			{ 410, 2 },
+
+			{ 424, 1 },
+			{ 425, 1 },
+			{ 426, 1 },
+
+			{ 612, 2 },
```
Comment (Arthri) [inline TShockAPI/Bouncer.cs]:
```suggestion
			{ 612, 0.7f },
```
Source: `Player.ItemCheck_Shoot()`, `float ai4 = (Main.rand.NextFloat() - 0.5f) * ((float)Math.PI / 4f);`
Increments every tick by `0.01`, has a lifetime of 60 for a total for +0.3. `Projectile.AI()`, `this.ai[1] += 0.01f;`

File: TShockAPI/Bouncer.cs [hunk 8]
```diff
@@ -2496,5 +2511,45 @@ internal static int GetMaxPlaceStyle(int tileID)
...
+			{ 756, 0 },
+			{ 522, 0 }
+		};
+		private Dictionary<short, short> Projectile_MaxValuesAI2 = new Dictionary<short, short> {
+			{ 405, 2 },
+			{ 410, 2 },
+
+			{ 424, 1 },
+			{ 425, 1 },
+			{ 426, 1 },
```
Comment (Arthri) [inline TShockAPI/Bouncer.cs]:
```suggestion
			{ 424, 0.8f },
			{ 425, 0.8f },
			{ 426, 0.8f },
```
Source: `Player.ItemCheck_Shoot`, `Projectile.NewProjectile(..., 0.5f + (float)Main.rand.NextDouble() * 0.3f);`

File: TShockAPI/Bouncer.cs [hunk 9]
```diff
@@ -2496,5 +2511,45 @@ internal static int GetMaxPlaceStyle(int tileID)
...
+			{ 950, 0 }
+		};
+
+		private Dictionary<short, short> Projectile_MinValuesAI2 = new Dictionary<short, short> {
+			{ 405, 0 },
+			{ 410, 0 },
+
+			{ 424, 0 },
+			{ 425, 0 },
+			{ 426, 0 },
```
Comment (Arthri) [inline TShockAPI/Bouncer.cs]:
```suggestion
			{ 424, 0.5f },
			{ 425, 0.5f },
			{ 426, 0.5f },
```
Source: `Player.ItemCheck_Shoot`, `Projectile.NewProjectile(..., 0.5f + (float)Main.rand.NextDouble() * 0.3f);`

File: TShockAPI/Bouncer.cs [hunk 10]
```diff
@@ -2496,5 +2511,45 @@ internal static int GetMaxPlaceStyle(int tileID)
...
+
+			{ 612, 0 },
+			{ 953, 0 },
```
Comment (Arthri) [inline TShockAPI/Bouncer.cs]:
```suggestion
			{ 953, 0.85f },
```
Source: `Projectile.Kill()`, `..., 0.85f + Main.rand.NextFloat() * 1.15f);`

File: TShockAPI/Bouncer.cs [hunk 11]
```diff
@@ -2496,5 +2511,45 @@ internal static int GetMaxPlaceStyle(int tileID)
...
+
+			{ 756, 0 },
```
Comment (Arthri) [inline TShockAPI/Bouncer.cs]:
```suggestion
			{ 756, 0.5f },
```
Source: `Player.ItemCheck_Shoot`, `Projectile.NewProjectile(..., Main.rand.NextFloat() * 0.5f + 0.5f);`

File: TShockAPI/Bouncer.cs [hunk 12]
```diff
@@ -2496,5 +2511,45 @@ internal static int GetMaxPlaceStyle(int tileID)
...
+
+			{ 424, 1 },
+			{ 425, 1 },
+			{ 426, 1 },
+
+			{ 612, 2 },
+			{ 953, 2 },
+
+			{ 756, 1 },
+			{ 522, 0 }
```
Comment (Arthri) [inline TShockAPI/Bouncer.cs]:
```suggestion
			{ 522, 40f }
```
Source: `Projectile.AI()`, `if (this.ai[1] > 40f) { Kill(); }`

**Review by Arthri — CHANGES_REQUESTED** [review]

File: TShockAPI/Bouncer.cs (hunk 1 above)
Comment (AgaSpace) [reply TShockAPI/Bouncer.cs]: Yes, that's much better.

**Comment by AgaSpace** [comment]
The condition that checks the projectile for modified AI does not work. In fact, it works, but always. (Responds to all projectiles)

**Review by Arthri — APPROVED** [review]

**Pull request merged**